### DISCUSSION TITLE
`static_vector` preview is available

### DISCUSSION THREAD

**Discussion by slavenf** [slavenf/sfl-library#10] [Show and tell]
@ZehMatt If you are interested, `static_vector` preview is available in `develop` branch. I haven't written documentation yet, but I will do it very soon and then everything will be ready for new release.

I took [`P0843r2`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0843r2.html) as a design guide. 

I haven't done it yet, but I will add the following member functions:
* `static_vector(size_type, default_init_t)`
* `void resize(size_type count, default_init_t)`

Those member functions are proposed by P0843r2 and are already available in `boost::container::static_vector`.

**Comment by slavenf** [comment]
There is newer version [`p0843r10`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p0843r10.html), but I started working on my `static_vector` long time ago. I will probably take some extra time to review that version of document and see if it is possible to implement that.

**Comment by ZehMatt** [comment]
Cool, will check it out once I have a bit of time.

**Comment by ZehMatt** [comment]
I've finally added the library to OpenLoco, the first use of it was added here https://github.com/OpenLoco/OpenLoco/pull/2436, seems to be working nicely so far.

**Comment by slavenf** [comment]
Great!!! I'm glad that it works without problems. I'm also happy that my library is used in such great game. I've checked the code you refactored and now it definitely looks easier to read and maintain.

If there are no suggestions or complains, I will merge this into main branch. 

`sfl::static_vector` does not throw exceptions if attempting to insert more than `N` elements. In debug mode such attempts are caught with `assert` macro. In release mode it is undefined behavior (i.e. crash sooner or later). If needed, exception support can be added later in backward compatible manner (I'm not 100% sure about this, but I think it is possible). I prefer using assertions because using exceptions is somewhat weird. My reasoning is the same as is in [p0843r2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0843r2.html). Yeah, there is a newer version of this paper [p0843r10](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p0843r10.html), but boost library and etlcpp library are more like p0843r2 and `sfl::static_vector` is more similar to boost and etlcpp.

I am currently working on static flat maps/sets and it will be available soon.

**Reply by ZehMatt** [reply]
I think the assertion is fine, it should be up to the user to not overfill the available storage. As far I'm concerned it looks ready but I will definitely let you know if I run into any issues. There are more places in the OpenLoco code that can make use of it, now with the library in I will definitely make more use of it, but it served a good starting point. Also static maps/sets sounds really nice too, given its fixed storage it should also open the opportunity to have constexpr initialization which is great making compile time lookup tables. Incredible work, thanks again.

**Reply by slavenf** [reply]
Thanks, I'm glad we are thinking the same. Of course, if you run into any issuse, please report it.
